source/common/listener_manager/listener_impl.cc

```diff
@@ -509,12 +509,24 @@ void ListenerImpl::buildInternalListener(const envoy::config::listener::v3::List
         parent_.server_.singletonManager().getTyped<Network::InternalListenerRegistry>(
             "internal_listener_registry_singleton");
     if (internal_listener_registry == nullptr) {
-      throw EnvoyException(fmt::format(
-          "error adding listener named '{}': internal listener registry is not initialized.",
-          name_));
+      // The internal listener registry may be uninitialized when in Validate mode.
+      // Hence we check the configuration directly to ensure the bootstrap extension
+      // "internal_listener" is present.
+      if (absl::c_none_of(
+              listener_factory_context_->serverFactoryContext().bootstrap().bootstrap_extensions(),
+              [=](const auto extension) {
+                return extension.typed_config().type_url() ==
+                       "type.googleapis.com/"
+                       "envoy.extensions.bootstrap.internal_listener.v3.InternalListener";
+              })) {
+        throw EnvoyException(
+            fmt::format("error adding listener named '{}': internal_listener is mandatory", name_));
```

Comment:
InternalListener bootstrap extension is mandatory

Reply:
Done